In [2]:
import pandas as pd
file_path_excel = R'国家认同数据清理后.xlsx'
df_excel = pd.read_excel(file_path_excel)
# 创建民族、年级和政治面貌的交互表
cross_tab = pd.crosstab(index=[df_excel['民族'], df_excel['年级']], 
                        columns=df_excel['政治面貌'], 
                        margins=True,normalize=True)  # All 行和列的总和

cross_tab


政治面貌           1         2         3         4       All
民族  年级                                                  
1   0   0.000000  0.000000  0.001284  0.000000  0.001284
    1   0.119384  0.007702  0.010270  0.002567  0.139923
    2   0.116816  0.019255  0.012837  0.002567  0.151476
    3   0.039795  0.028241  0.007702  0.005135  0.080873
    4   0.028241  0.014121  0.002567  0.000000  0.044929
...          ...       ...       ...       ...       ...
36  1   0.001284  0.000000  0.000000  0.000000  0.001284
47  2   0.001284  0.000000  0.000000  0.000000  0.001284
48  1   0.000000  0.001284  0.000000  0.000000  0.001284
97  2   0.001284  0.000000  0.000000  0.000000  0.001284
All     0.736842  0.132221  0.110398  0.020539  1.000000

[78 rows x 5 columns]

In [3]:
from scipy import stats
# 假设我们关注的是第二年级的学生（假设年级用数字“2”表示）
specific_grade = 2

# 筛选出特定年级的数据
df_specific_grade = df_excel[df_excel['年级'] == specific_grade]

# 创建政治面貌和会打多少分的交互表
cross_tab_specific_grade = pd.crosstab(df_specific_grade['会打多少分'], df_specific_grade['政治面貌'])

# 进行卡方检验
chi2_grade, p_grade, dof_grade, expected_grade = stats.chi2_contingency(cross_tab_specific_grade)

print(chi2_grade,p_grade)
# chi2_grade, p_grade, dof_grade, expected_grade[:10]  # 展示期望频次的前10个结果，避免过长的输出

20.269905911839633 0.06214943009111473


In [4]:
import pandas as pd
file_path_excel = R'国家认同数据清理后.xlsx'

In [5]:
df_excel = pd.read_excel(file_path_excel)
# 创建年级、政治面貌和会打多少分的交互表
cross_tab = pd.crosstab(index=[df_excel['年级'], df_excel['会打多少分']], 
                        columns=df_excel['政治面貌'], 
                        margins=True, normalize=True)  # All 行和列的总和

cross_tab

政治面貌              1         2         3         4       All
年级  会打多少分                                                  
0   3      0.005135  0.000000  0.000000  0.000000  0.005135
    4      0.008986  0.000000  0.000000  0.000000  0.008986
    5      0.001284  0.000000  0.001284  0.000000  0.002567
1   1      0.006418  0.000000  0.001284  0.000000  0.007702
    2      0.010270  0.000000  0.001284  0.000000  0.011553
    3      0.032092  0.005135  0.003851  0.003851  0.044929
    4      0.151476  0.007702  0.016688  0.000000  0.175866
    5      0.091142  0.002567  0.010270  0.000000  0.103979
2   1      0.011553  0.001284  0.000000  0.000000  0.012837
    2      0.012837  0.001284  0.001284  0.000000  0.015404
    3      0.043646  0.011553  0.016688  0.001284  0.073171
    4      0.110398  0.007702  0.010270  0.003851  0.132221
    5      0.066752  0.012837  0.005135  0.001284  0.086008
3   1      0.003851  0.002567  0.002567  0.000000  0.008986
    2      0.002567  0.001284  0.000000  0.000000  0.003851
    3      0.025674  0.010270  0.006418  0.001284  0.043646
    4      0.048780  0.033376  0.015404  0.005135  0.102696
    5      0.030809  0.008986  0.003851  0.003851  0.047497
4   1      0.000000  0.002567  0.000000  0.000000  0.002567
    2      0.001284  0.000000  0.001284  0.000000  0.002567
    3      0.010270  0.005135  0.005135  0.000000  0.020539
    4      0.037227  0.007702  0.005135  0.000000  0.050064
    5      0.024390  0.010270  0.002567  0.000000  0.037227
All        0.736842  0.132221  0.110398  0.020539  1.000000

In [6]:
from scipy import stats
# 假设我们关注的是第二年级的学生（假设年级用数字“2”表示）
specific_grade = 1

# 筛选出特定年级的数据
df_specific_grade = df_excel[df_excel['年级'] == specific_grade]

# 创建政治面貌和会打多少分的交互表
cross_tab_specific_grade = pd.crosstab(df_specific_grade['会打多少分'], df_specific_grade['政治面貌'])

# 进行卡方检验
chi2_grade, p_grade, dof_grade, expected_grade = stats.chi2_contingency(cross_tab_specific_grade)

print(chi2_grade,p_grade)
cross_tab_specific_grade

26.390294602838058 0.009448085692166378


政治面貌     1  2   3  4
会打多少分               
1        5  0   1  0
2        8  0   1  0
3       25  4   3  3
4      118  6  13  0
5       71  2   8  0

In [7]:
# 创建政治面貌和会打多少分的交互表
cross_tab_specific_grade = pd.crosstab(df_excel['会打多少分'], df_excel['政治面貌'])

# 进行卡方检验
chi2_grade, p_grade, dof_grade, expected_grade = stats.chi2_contingency(cross_tab_specific_grade)
print(chi2_grade, p_grade)

16.22193387826521 0.1812812450681659


In [8]:
数据表 = pd.read_spss(R'国家认同数据清理后.sav')

In [9]:
数据表.columns

Index(['问卷编号', '调查员', '民族', '政治面貌', '年级', '您觉得自己是个典型的中国人吗',
       '与世界其他国家的人相比中国人有自己的特点吗', '为公民，最基本的要求是爱自己的国家', '不管是汉族还是少数民族，都是中华民族的一员',
       '您是否为为中华民族的历史文化而骄傲', '个人的前途与命运是与中华民族的命运息息相关', '您是否了解中华民族的传统节日',
       '您是否了解中国历史、地理、政治等', '老实说，您觉得中国怎么样', '您认为中国有多少值得自豪的地方',
       '您认为世界有多少比例的人尊重中国', '对您而言作为一名中国人有多重要', '会以中国人自豪吗', '会隐瞒身份吗', '会打多少分',
       '国歌升起', '世博会', '中国传统文化', '发展信心', '你会为中国运动员呐喊助威', '遇到灾难时中国人应该伸出援手',
       '你愿意加入其他国籍吗', '中国人要为祖国统一奋斗吗'],
      dtype='object')

In [10]:
import pandas as pd
file_path_spss = R'国家认同数据清理后.sav'

In [11]:
df_sav = pd.read_spss(file_path_spss)
# 创建年级、政治面貌和会以中国人自豪吗的交互表
cross_tab = pd.crosstab(index=[df_sav['政治面貌'], df_sav['会以中国人自豪吗']], 
                        columns=df_sav['年级'], 
                        margins=True, normalize=True)  # All 行和列的总和

cross_tab

年级                   大一        大三        大二        大四        预科       All
政治面貌 会以中国人自豪吗                                                            
党员   不会        0.000000  0.002567  0.003851  0.000000  0.000000  0.006418
     不太会       0.001284  0.007702  0.000000  0.003851  0.000000  0.012837
     会         0.010270  0.023107  0.012837  0.010270  0.000000  0.056483
     无所谓       0.001284  0.005135  0.012837  0.001284  0.000000  0.020539
     有些        0.002567  0.017972  0.005135  0.010270  0.000000  0.035944
其他   不会        0.000000  0.000000  0.001284  0.000000  0.000000  0.001284
     不太会       0.001284  0.000000  0.000000  0.000000  0.000000  0.001284
     会         0.001284  0.003851  0.001284  0.000000  0.000000  0.006418
     无所谓       0.001284  0.002567  0.002567  0.000000  0.000000  0.006418
     有些        0.000000  0.003851  0.001284  0.000000  0.000000  0.005135
团员   不会        0.003851  0.002567  0.005135  0.002567  0.001284  0.015404
     不太会       0.017972  0.003851  0.015404  0.003851  0.003851  0.044929
     会         0.164313  0.055199  0.107831  0.041078  0.003851  0.372272
     无所谓       0.029525  0.017972  0.038511  0.010270  0.005135  0.101412
     有些        0.075738  0.032092  0.078306  0.015404  0.001284  0.202824
群众   不太会       0.001284  0.002567  0.000000  0.000000  0.000000  0.003851
     会         0.019255  0.006418  0.012837  0.006418  0.001284  0.046213
     无所谓       0.005135  0.010270  0.014121  0.003851  0.000000  0.033376
     有些        0.007702  0.008986  0.006418  0.003851  0.000000  0.026958
All            0.344031  0.206675  0.319641  0.112965  0.016688  1.000000

In [12]:
import pyreadstat
import pandas as pd
from scipy import stats
# 读取.sav文件
df, meta = pyreadstat.read_sav('国家认同数据清理后.sav')
# 假设我们关注的是第三年级的学生（假设年级用数字“2”表示）
specific_grade = 3
# 筛选出特定年级的数据
df_specific_grade = df[df['年级'] == specific_grade]
# 创建政治面貌和会以中国人自豪吗的交互表
cross_tab_specific_grade = pd.crosstab(df_specific_grade['政治面貌'], df_specific_grade['会以中国人自豪吗'])
# 进行卡方检验
chi2_grade, p_grade, dof_grade, expected_grade = stats.chi2_contingency(cross_tab_specific_grade)

print(chi2_grade, p_grade)
print(cross_tab_specific_grade)

16.610955615265958 0.16482706405625666
会以中国人自豪吗  1.0  2.0  3.0  4.0  5.0
政治面貌                             
1.0         2    3   14   25   43
2.0         2    6    4   14   18
3.0         0    2    8    7    5
4.0         0    0    2    3    3


这个卡方检验的结果显示，特定年级学生的政治面貌和是否以中国人自豪之间的关系并不具有统计学意义。具体来说，卡方值为16.61，p值为0.165。根据通常的显著性水平（通常为0.05），p值大于0.05意味着我们无法拒绝原假设，即观察到的结果可能是由随机因素引起的，而非真正的关联。因此，根据这个p值，我们无法得出特定年级学生的政治面貌和是否以中国人自豪之间存在显著关联的结论。
这意味着在这个特定年级的学生中，政治面貌和是否以中国人自豪之间的关系可能是随机的，而非由于学生的政治面貌导致的。这个结果可能需要更多的数据或者其他分析方法来进一步确认是否存在真正的关联。

In [13]:
import mytools
mytools.两个无序类别变量的统计分析(df_specific_grade,'政治面貌','会以中国人自豪吗')

tau_y系数: 0.0264 极弱相关或无相关
-  -  -  --  --  --
1  2  3  14  25  43
2  2  6   4  14  18
3  0  2   8   7   5
4  0  0   2   3   3
-  -  -  --  --  --
卡方值： 16.61, p值： 0.1648,自由度:12。
接受虚无假设
